### PR TITLE
test: complete node harness framework semantics

### LINKED ISSUE #256: node-test-harness.js: complete test-framework semantics for the library clusters (#233–236)
Deferred from #253 (harness for #232). The in-process harness is proven on lodash (QUnit, flat modules, explicit `QUnit.load()`+`QUnit.start()`) at 6,794 assertions cross-checked. Codex review surfaced three further semantics that the lodash proof does not exercise and that belong with the clusters that will actually consume them (mocha/jest/tape via the describe/it spine, or more QUnit suites) rather than this proof PR. Each is a real gap, tracked here:

1. **QUnit `autostart`** (`scripts/node-test-harness.js`, `QUnit.load`). `QUnit.load()` is currently a no-op; a suite that relies on default autostart (registers tests without calling `QUnit.start()`) never runs, so the runner sees no `PASS/FAIL/TOTAL` line. lodash calls `start()` explicitly, so it's unaffected. Fix: run the suite after synchronous registration when `config.autostart` is true and `start()` wasn't called.

2. **mocha/tape `done`-callback async tests** (describe/it TAP runner). `it('x', function (done) { setTimeout(done, 0) })` is invoked with no argument and its `undefined` return is treated as completion, so the test is reported passed before the async work finishes. Fix: when the test/hook fn declares a parameter, pass a `done` callback and await its call (with a timeout guard), like the QUnit `assert.async()` path.

3. **Nested QUnit module hook inheritance.** A `QUnit.module()` nested inside another module's callback creates an independent module; `runTest` only consults `testObj.module.hooks`, so a child test doesn't inherit the parent module's `before`/`beforeEach`/`afterEach`/`after`. Modern QUnit inherits outer-module hooks. Fix: walk the parent chain for hooks (and run before/after at the right nesting boundaries).

These are the reusable-spine refinements the epic anticipated deferring; they should land with (or just ahead of) the first cluster that exercises them.

### PR DESCRIPTION
**Pull request by pmatos**
## Summary

- start QUnit suites through default autostart and `QUnit.load()` after registration
- preserve nested QUnit module ancestry, definition order, inherited per-test hooks, and module hook boundaries
- await mocha/tape-style `done` callbacks for tests and hooks with error propagation and a cancellable timeout
- add deterministic harness fixtures for all three semantics and document the supported behavior

No engine globals, Rust sources, dependencies, test262 files, or pass baselines change.

## Validation

- `./scripts/run-harness-selftest.sh --no-build` — green
- `node --check scripts/node-test-harness.js` — green
- `./scripts/lint.sh` — green (`cargo fmt --check`, clippy with warnings denied)
- `cargo build --release` — green
- `cargo test --release` — 291 unit tests + smoke oracle green
- `uv run python scripts/run-custom-tests.py` — 6/6 green
- targeted ExpectedArgumentCount/function-length test262 — 8/8 green
- lodash — jsse 6,794/6,794; Node CommonJS oracle 6,794/6,794
- full test262 — 99,546/99,568 with 22 pre-existing/environmental Intl402 DateTimeFormat/Temporal baseline failures; this scripts-only diff has no engine or dependency changes

The stock lodash Node command is affected on this host by an unrelated `/tmp/package.json` with `"type": "module"`; reading the identical final bundle as CommonJS validates the Node oracle without mutating that external project. Details are recorded on #256.

Closes #256

### CODE REVIEW

File: scripts/node-test-harness.js [hunk 1]
```diff
@@ -1176,9 +1236,37 @@
...
+        } catch (e) {
+          done(e);
+        }
+      });
+    }
+
     async function runHooks(hooks, ctx) {
       for (var i = 0; i < hooks.length; i++) {
-        await Promise.resolve(hooks[i].call(ctx));
+        await invokeRunnable(hooks[i], ctx);
```
[the remark this answers is not in the document]
Comment (pmatos) [reply scripts/node-test-harness.js:1280]: Fixed in `70364a4`.

Changed: `runSuite()` now contains suite-level `before all` / `after all` hook failures (thrown, `done(error)`, or timed out) the same way per-test `beforeEach`/`afterEach` already are in `runOneTest`. A failing hook is recorded as a single `not ok`; a failed `before all` skips its suite's children (their fixture state is unreliable) but lets sibling suites still run; `after all` runs regardless, for cleanup. This removes the rejection path through `runAll` → the top-level harness catch that collapsed every count to `PASS: 0  FAIL: 1  TOTAL: 1` and aborted sibling suites. A shared `reportFailure()` helper is now used by both failing tests and failing hooks so they report identically.

Validation: new `scripts/harness-fixtures/tap-suite-hook-failure.fixture.js` (done(error) `before`/`after` + a synchronously-throwing `before all`, siblings proven to keep running) asserts `PASS: 2  FAIL: 3  TOTAL: 5`; `./scripts/run-harness-selftest.sh` green across all 5 fixtures; `node --check scripts/node-test-harness.js` and `cargo build --release` green.

File: scripts/node-test-harness.js [hunk 2]
```diff
@@ -966,23 +999,40 @@
...
+          common++;
         }
-        await runModuleHook(mod, "after");
+        for (var close = activeModules.length - 1; close >= common; close--) {
+          await runModuleHook(activeModules[close], "after");
+        }
+        activeModules.length = common;
+        for (var open = common; open < nextModules.length; open++) {
+          await runModuleHook(nextModules[open], "before");
+          activeModules.push(nextModules[open]);
```
[the remark this answers is not in the document]
Comment (pmatos) [reply scripts/node-test-harness.js]: Fixed in `5da663c`.

Changed: the QUnit adapter now seeds each module's `sharedEnv` from its parent's when the module is opened in `runAll()` (shallow copy, this module layered on top), so a nested module's own `before`/`after` hooks and its tests inherit `this` state set by ancestor `before` hooks — matching QUnit's nested testEnvironment inheritance. `runTest()`'s per-test `Object.assign({}, module.sharedEnv)` therefore now carries the full ancestor chain.

Validation: confirmed the expected semantics against real qunit@2.26.0 (a nested child test sees the parent `before`'s `this.fixture`; a child `before` sees parent state and layers on it; a child `before` overrides an inherited value). New `scripts/harness-fixtures/qunit-nested-hook-state.fixture.js` asserts `PASS: 4  FAIL: 0  TOTAL: 4` (it is `PASS: 1  FAIL: 3` on the pre-fix harness); `./scripts/run-harness-selftest.sh` is green across all 6 fixtures, including the pre-existing nested `qunit.fixture.js` (unchanged at `PASS: 23  FAIL: 1  TOTAL: 24`).

No risk to the lodash cross-check: the change only affects the nested-module branch and is a no-op for flat, hook-less modules — all 307 `QUnit.module` calls in lodash 4.17.21 are the single-arg flat form, so lodash never reaches the new code path and its leaf `testEnv` resolves identically.

File: scripts/node-test-harness.js [hunk 3]
```diff
@@ -1176,12 +1247,54 @@
...
+          settled = true;
+          reject(
+            new Error(
+              "done callback timed out after " + ASYNC_TIMEOUT_MS + "ms"
+            )
+          );
+        }, ASYNC_TIMEOUT_MS);
+        function done(error) {
+          if (settled) return;
+          settled = true;
```
[the remark this answers is not in the document]
Comment (pmatos) [reply scripts/node-test-harness.js:1265]: **Deferred (diminishing returns)** — the claim is correct, but this is a pathological edge case on the `describe/it` TAP path, which no currently-wired library exercises (the wired libraries are QUnit-based) and which has no Node oracle (the harness is inert on Node — there the library's own framework runs), so it is a latent robustness gap rather than an active count divergence. A faithful mocha-matching fix is also non-trivial: a synchronous `done(); done(err)` can't be turned into a failure once the promise has already resolved on the first call, an asynchronous late duplicate arrives after the test was already reported `ok`, and the `settled` guard's timeout-vs-done idempotency must be preserved — not worth the churn/risk in this scripts-only PR. Tracked as #272 (https://github.com/pmatos/jsse/issues/272).

_This assessment was made by two independent AI reviewers (Claude Opus 4.8 and Codex GPT-5.5), both DEFER with high confidence. If you disagree, please reply and we'll re-evaluate._
